File: Xamarin/mySkatemap.TrackRecorderApp/Infrastructure/Location/FusedLocationProvider.cs
using Android.App;
using Android.Gms.Location;

namespace mySkatemap.Apps.TrackRecorder.Infrastructure.Location
{
	internal sealed class FusedLocationProvider : LocationProvider
	{
		private readonly FusedLocationProviderClient fusedLocationProviderClient;

		private readonly LocationCallback locationCallback;

		public FusedLocationProvider()
		{
			this.fusedLocationProviderClient = LocationServices.GetFusedLocationProviderClient(Application.Context);

			this.locationCallback = new LocationCallback();
			this.locationCallback.LocationResult += (sender, args) =>
			{
				foreach (var location in args.Result.Locations)
				{
					this.PostLocationUpdate(new Location
					{
						Latitude = location.Latitude
					});
				}
			};
		}

		public override void StartLocationUpdates()
		{
			if (this.HasRequestedLocationUpdates)
			{
				return;
			}

			var locationRequest = new LocationRequest();
			locationRequest.SetPriority(LocationRequest.PriorityHighAccuracy);
			locationRequest.SetInterval(8000);
			locationRequest.SetFastestInterval(4000);

			this.fusedLocationProviderClient.RequestLocationUpdatesAsync(locationRequest, this.locationCallback);

			this.HasRequestedLocationUpdates = true;
		}

		public override void StopLocationUpdates()
		{
			if (!this.HasRequestedLocationUpdates)
			{
				return;
			}

			this.fusedLocationProviderClient.RemoveLocationUpdatesAsync(this.locationCallback);

			this.HasRequestedLocationUpdates = false;
		}

		public override void RequestLocation()
		{
			var lastLocation = this.fusedLocationProviderClient.GetLastLocationAsync().Result;

			this.PostLocationUpdate(new Location
			{
				Latitude = lastLocation.Latitude
			});
		}
	}
}

File: Android/app/src/main/java/com/janhafner/myskatemap/apps/trackrecorder/location/provider/FusedLocationProvider.kt
package com.janhafner.myskatemap.apps.trackrecorder.location.provider

import android.annotation.SuppressLint
import android.content.Context
import com.google.android.gms.location.*
import com.janhafner.myskatemap.apps.trackrecorder.location.Location
import com.janhafner.myskatemap.apps.trackrecorder.toLocation

internal final class FusedLocationProvider(context: Context): LocationProvider() {
    private val fusedLocationProviderClient: FusedLocationProviderClient

    private val locationCallback: LocationCallback = object: LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            val self = this@FusedLocationProvider

            for(sourceLocation in locationResult.locations) {
                val sequenceNumber = self.generateSequenceNumber()

                val location = sourceLocation.toLocation(sequenceNumber)

                self.postLocationUpdate(location)
            }
        }
    }

    private val locationRequest: LocationRequest = LocationRequest()

    init {
        locationRequest.interval = 8000
        locationRequest.fastestInterval = 4000
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY

        this.fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
    }

    @SuppressLint("MissingPermission")
    public override fun startLocationUpdates() {
        if (this.isActive) {
            throw IllegalStateException()
        }

        this.fusedLocationProviderClient.requestLocationUpdates(this.locationRequest, this.locationCallback, null)

        this.isActive = true
    }

    public override fun stopLocationUpdates() {
        if (!this.isActive) {
            throw IllegalStateException()
        }

        this.fusedLocationProviderClient.removeLocationUpdates(this.locationCallback)

        this.isActive = false
    }

    @SuppressLint("MissingPermission")
    public override fun getCurrentLocation(): Location? {
        return this.fusedLocationProviderClient.lastLocation.result.toLocation(-1)
    }
}